docs/content/sips/001-spin-deploy.md

```diff
@@ -0,0 +1,47 @@
+title = "SIP 001 - Spin Deploy"
+template = "main"
+date = ""
+
+Summary: A Spin CLI command that will deploy a Spin application to the cloud
+
+Owner: michelle@fermyon.com
+
+Created: March 18, 2022
+
+Updated: March 31, 2022
+
+## Background
+
+A Spin application — (currently described as a spin.toml file) — is a collection of one or more components, each invoked as a result of an event generated by a trigger. The Spin CLI currently enables running a Spin application locally via the `spin up` command. The Spin CLI also offers a set of commands enabling users to package and push a Spin application to a [Bindle](https://github.com/deislabs/bindle) registry.
+
+[Hippo](https://github.com/deislabs/hippo) is a self-hosted, open source, Web Assembly PaaS that allows users to deploy and scale Web Assembly applications at scale. Hippo takes a reference to a bindle, deploys it to the cloud and makes it publicly available. Hippo currently only works with Wagi applications and can currently deploy Wagi applications on a Nomad cluster with Traefik for routing. Additionally, Hippo has a concept called “channels” enabling users to track different version categories of an application (ex. alpha, beta, stable, pre-releases, minor versions, etc.).
+
+A desired feature is to be able to deploy a Spin application to the cloud using Hippo so that it is accessible via a publicly available domain or IP address.
+
+## Proposal
+
+Create a `spin deploy` command. This command packages a Spin application as a bindle, pushes it to a bindle registry and instructs Hippo to deploy the application to the cloud.
+
+This command needs the following configuration to be passed via either flags or environment variables or config file to deploy the application properly and to the right place:
+
+- Bindle server information
+  - Environment variable: BINDLE_SERVER_URL
+  - Flag: --bindle-server
+- Hippo URL information
+  - Environment variable: HIPPO_URL
+  - Flag: --hippo-url
+- Hippo authentication information
+  - Environment variable: HIPPOTOKEN
+  - Flag: --hippo-token
+  - _Note: currently the Hippo auth token expires in 30 minutes and there is no way to refresh the token. We may need to support basic auth in the first iteration as token support will require updates to Hippo._
```

Comment:
A refresh token API is [expected to land the next Hippo release](https://github.com/deislabs/hippo/issues/545), so perhaps it'd be worth going through with implementing client token support. Implementing token refreshes would only be a few lines of code once 0.6.0 lands.

Have a look at hippo's UI for a good implementation of session management: https://github.com/deislabs/hippo/blob/main/src/Web/ClientApp/src/app/_services/session.service.ts

You could implement "forced login" from two directions:

1. Implementing an HTTP interceptor, sniffing incoming responses from Hippo. If a 401 response is returned, [ask the user to log back in](https://github.com/deislabs/hippo/blob/2c69a2c2eeb2b84ddbd4b3ba5a6d9bd4a95a979a/src/Web/ClientApp/src/app/_helpers/auth.interceptor.ts#L18-L30).
2. Check the token's expiry date before sending a request. If it has expired, [force the user to log back in](https://github.com/deislabs/hippo/blob/2c69a2c2eeb2b84ddbd4b3ba5a6d9bd4a95a979a/src/Web/ClientApp/src/app/_helpers/auth.guard.ts#L19-L22). Eventually, this could be replaced with the refresh token API.

If it helps, here's hippo-cli's client implementation: https://github.com/deislabs/hippo-cli/blob/main/src/hippo/client.rs